tests/test_query_engine.py

```diff
@@ -616,15 +615,189 @@ def test_categorise(database, setup_test_database):
+                "tallish": (height_value > 175) & (height_value != 180),
+                "short": height_value <= 175,
+            },
+            "missing",
+            [
+                dict(patient_id=1, height_group="missing"),
+                dict(patient_id=2, height_group="tallish"),
+                dict(patient_id=3, height_group="missing"),
+                dict(patient_id=4, height_group="short"),
+            ],
+        ),
+    ],
+    ids=[
+        "test simple and on two conditions",
+        "test shorthand between conditions",
+        "test simple or on two conditions",
+        "test a not-equals condition",
+    ],
+)
+def test_categorise_single_combined_conditions(
+    database, setup_test_database, categories, default, expected
+):
+    input_data = [
+        RegistrationHistory(PatientId=1, StpId="STP1"),
+        RegistrationHistory(PatientId=2, StpId="STP2"),
+        RegistrationHistory(PatientId=3, StpId="STP2"),
+        RegistrationHistory(PatientId=4, StpId="STP2"),
+        Patients(PatientId=1, Height=180),
+        Patients(PatientId=2, Height=200.5),
+        Patients(PatientId=3),
+        Patients(PatientId=4, Height=145),
+    ]
+    setup_test_database(input_data)
+
+    class Cohort:
+        _height = table("patients").first_by("patient_id").get("height")
+        _height_categories = categories(_height)
+        height_group = categorise(_height_categories, default=default)
+
+    result = list(extract(Cohort, MockBackend, database))
+    assert result == expected
+
+
+@pytest.mark.integration
+def test_categorise_multiple_values(database, setup_test_database):
+    """Test that categories can combine conditions that use different source values"""
+    input_data = [
+        RegistrationHistory(PatientId=1, StpId="STP1"),
+        RegistrationHistory(PatientId=2, StpId="STP2"),
+        RegistrationHistory(PatientId=3, StpId="STP2"),
+        Patients(PatientId=1, Height=200),
+        Patients(PatientId=2, Height=150),
+        Patients(PatientId=3, Height=160),
+        Events(PatientId=1, EventCode="abc"),
+        Events(PatientId=2, EventCode="xyz"),
+        Events(PatientId=3, EventCode="abc"),
+    ]
+    setup_test_database(input_data)
+
+    class Cohort:
+        _height = table("patients").first_by("patient_id").get("height")
+        _code = (
+            table("clinical_events")
+            .filter(code="abc")
+            .first_by("patient_id")
+            .get("code")
+        )
+        _height_with_codes_categories = {
+            "short": (_height < 190) & (_code == "abc"),
+            "tall": (_height > 190) & (_code == "abc"),
+        }
+        height_group = categorise(_height_with_codes_categories, default="missing")
+
+    result = extract(Cohort, MockBackend, database)
+    assert result == [
+        dict(patient_id=1, height_group="tall"),
+        dict(patient_id=2, height_group="missing"),
+        dict(patient_id=3, height_group="short"),
+    ]
```

Comment:
A general comment on testing, with this as a specific example. I think we'd benefit from trying to slim down each test case as much as possible -- reducing the test data and cohort definition to the absolute minimum. This makes the purpose of the tests easier to understand, more clearly points at the semantics of the feature and reduces the chance that we'll need to modify the test in future.

Here, for example, I think we only need a single STP and I don't think the `filter` in the code handling is necessary.

I know it's tempting to grab what we already have to hand and reuse it, but I think the extra effort will pay off in the long term.

Reply:
I think the filter in the code is a leftover from me initially trying to use an `exists` on it.  I'll simplify it